addons/mail/static/src/core/common/message_actions.js

```diff
@@ -68,17 +68,46 @@ registerMessageAction("reaction", {
     sequence: 10,
 });
 registerMessageAction("reply-to", {
-    condition: (component) => component.props.message.canReplyTo(component.props.thread),
+    condition: (component) => {
+        const { message, thread } = component.props;
+        return (
+            message.canReplyTo(thread) ||
+            (!["discuss.channel", "mail.box"].includes(thread.model) &&
+                message.isNote &&
+                !message.isSelfAuthored)
+        );
+    },
     icon: "fa fa-reply",
     iconLarge: "fa fa-lg fa-reply",
     name: _t("Reply"),
     onSelected: (component) => {
         const message = toRaw(component.props.message);
         const thread = toRaw(component.props.thread);
-        if (message.eq(thread.composer.replyToMessage)) {
-            thread.composer.replyToMessage = undefined;
+        const composer = thread.composer;
+        let currentText = composer.text || "";
+        const mentionText = `@${message.authorName} `;
+        if (message.eq(composer.replyToMessage)) {
+            composer.replyToMessage = undefined;
         } else {
-            thread.composer.replyToMessage = message;
+            if (["discuss.channel", "mail.box"].includes(thread.model)) {
+                composer.replyToMessage = message;
+            }
+            if (thread.model !== "discuss.channel") {
+                if (!message.isSelfAuthored && message.model !== "discuss.channel") {
+                    if (!composer.mentionedPartners.includes(message.author)) {
+                        composer.mentionedPartners.push(message.author);
+                    }
+                    if (!currentText.includes(mentionText)) {
+                        currentText = mentionText + currentText;
+                    }
+                }
+                composer.text = currentText;
+                composer.selection.start = currentText.length;
+                composer.selection.end = currentText.length;
+                if (component.env.inChatter?.composerType !== "note") {
+                    component.env.toggleComposer?.("note");
+                }
+            }
```

Comment:
Prefer early returns rather than lots of nested if/else: single if/else that have no extra indent and are short is fine, but here it's easier to read with:
```js
if () {
   ...
   return;
}
...
if () {
  ...
  return;
}
...
```